leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -116,24 +117,37 @@ private void tryToGainLeadership() {
 
     private void tryToGainLeadershipAsync() {
         try {
-            executor.submit(this::gainLeadershipBlocking);
+            executor.submit(this::gainLeadershipWithRetry);
         } catch (RejectedExecutionException e) {
             if (!isClosed) {
                 throw new IllegalStateException("failed to submit task but proxy not closed", e);
             }
         }
```

Comment:
I think submit can throw in two cases:
1- Executor service is shutdown; which should only happen when we close the proxy, and in that case we shouldn't retry.
2- ThreadPool is filled, and executor service is rejecting to queue a new task - first this should never happen (we should only submit this task once - so that would point a bug in the code), second even if that happens we should be fine, as that would mean we already have a task trying to gain leadership.

So I think we shouldn't retry in this case.